src/broadcast/scrobblingmanager.cpp

```diff
@@ -0,0 +1,215 @@
+    Deck *sourceDeck = qobject_cast<Deck*>(sender());    
+    if (sourceDeck == 0) {
+        qDebug() << "Didn't load track in a deck yet scrobbling "
+                    "received paused signal.";
+        return;
+    }
+    // QMutexLocker locker(&m_mutex);
+    bool allPaused = true;
+    TrackInfo *pausedTrackInfo;
+    for (TrackInfo *trackInfo : m_trackList) {
+        if (trackInfo->m_pTrack == pPausedTrack) {
+            pausedTrackInfo = trackInfo;
+            for (BaseTrackPlayer *player : trackInfo->m_players) {
+                BaseTrackPlayerImpl *playerImpl = 
+                    qobject_cast<BaseTrackPlayerImpl*>(player);
+                if (playerImpl == 0) {
+                    qDebug() << "Track player interface isn't a "
+                                "BaseTrackPlayerImpl";
+                    return;
+                }
+                if (!playerImpl->isTrackPaused())
+                    allPaused = false;
+            }
+            break;
+        }                   
+    }
+    if (allPaused)
+        pausedTrackInfo->m_trackInfo.pausePlayedTime();        
+}
+
+void ScrobblingManager::slotTrackResumed(TrackPointer pResumedTrack) {
+    //Extra functional decision to only track main decks.
+    Deck *sourceDeck = qobject_cast<Deck*>(sender());    
+    if (sourceDeck == 0) {
+        qDebug() << "Didn't load track in a deck yet scrobbling "
+                    "received resumed signal.";
+        return;
+    }    
+    if (isTrackAudible(pResumedTrack,sourceDeck)) {        
+        // QMutexLocker locker(&m_mutex);
+        for (TrackInfo *trackInfo : m_trackList) {
+            if (trackInfo->m_pTrack == pResumedTrack) {
+                trackInfo->m_trackInfo.resumePlayedTime();
+                break;
+            }
+        }  
+    }
+}
+
+void ScrobblingManager::slotLoadingTrack(TrackPointer pNewTrack, TrackPointer pOldTrack) {
+    //Extra functional decision to only track main decks.
+    Deck *sourceDeck = qobject_cast<Deck*>(sender());    
+    if (sourceDeck == 0) {
+        qDebug() << "Didn't load track in a deck yet scrobbling "
+                    "received loading signal.";
+        return;
+    }
+    if (pOldTrack) {
+        m_tracksToBeReset.append(TrackToBeReset(pOldTrack,sourceDeck));
+    }
+}
+
+void ScrobblingManager::slotNewTrackLoaded(TrackPointer pNewTrack) {
+    //Empty deck gives a null pointer.
+    if (!pNewTrack)
+        return;    
+    //Extra functional decision to only track main decks.        
+    Deck *newDeck = qobject_cast<Deck*>(sender());    
+    if (newDeck == 0) {
+        qDebug() << "Didn't load track in a deck yet scrobbling "
+                    "received loaded signal.";
+        return;
+    }
+    // QMutexLocker locker(&m_mutex);    
+    bool trackAlreadyAdded = false;
+    for (TrackInfo *trackInfo : m_trackList) {
+        if (trackInfo->m_pTrack == pNewTrack) {
+            trackInfo->m_players.append(newDeck);               
```

Comment:
Storing a pointer you not own is always kind of dangerous.
It is better to use a shared pointer, or the group string.

Reply:
Well, the player manager doesn't destroy the decks until it itself gets destroyed, and this object gets destroyed too. But I will use a group string because shared pointer implies modifying player manager.